src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs

```diff
@@ -121,38 +126,64 @@ protected override void FormatInlineCount(IList<string> commandClauses)
         }
 
         private IList<string> SelectPathSegmentColumns(
-            IList<string> columns, string path, IList<string> excludePaths = null)
+            IList<string> columns, string path, EntityCollection collection, IList<string> excludePaths = null)
         {
             if (string.IsNullOrEmpty(path))
             {
                 var resultColumns = columns.Where(x => !HasMultipleSegments(x)).ToList();
                 if (excludePaths != null)
-                    resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
-                        !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                {
+                    if (excludePaths.Count == 1 && excludePaths[0] == "*")
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !_session.Metadata.GetNavigationPropertyNames(collection.Name).Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                    }
+                    else
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                    }
+                }
+                   
                 return resultColumns;
             }
             else
             {
-                return columns.Where(x => HasMultipleSegments(x) && FormatFirstSegment(x) == FormatFirstSegment(path))
-                    .Select(x => FormatSkipSegments(x, 1)).ToList();
+                var resultColumns = FormatFirstSegment(path) == "*"
```

Comment:
`FormatFirstSegment(path)` should really be hoisted out into a local. `String.Split` will get called a lot otherwise.

Reply:
Same as above..